### ISSUE TITLE
SGLang rolling generation produces grammar-invalid trajectories where HF does not

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/MEDS_EIC_AR#171]
## Summary

Driven through the rolling-generation loop, the SGLang backend produces **grammar-invalid** trajectories where the HF backend produces 100%-valid ones from the same checkpoint. `tests/grammar/test_cli_sglang.py` fails its correctness assertion on **128/128** held-out samples.

This is **not** caused by #168/#169 — I reproduced it identically on their base (`feat/sglang-backend` @ `5649550`). Filing separately so it isn't mistaken for fallout from those PRs.

## Evidence

Hardware: DGX Spark (GB10, SM 12.1), `sglang==0.5.9`, `sgl-kernel==0.3.21+cu130`, `torch==2.9.1+cu129`.

To get past the context ceiling (#162) I ran the grammar fixture with `rolling_generation.rolling_context_size=13` (`max_seq_len=16`). Both backends in one pytest session, same fixture, same checkpoint:

```
tests/grammar/test_cli.py::test_grammar_cli_greedy_output_is_fully_grammar_valid   PASSED
tests/grammar/test_cli_sglang.py::test_sglang_greedy_output_is_fully_grammar_valid FAILED
```

So the `rolling_context_size=13` setting is not itself the cause — HF is fine with it.

The SGLang failures are uniform:

```
traj=0 subject=273: 5/15 valid (first invalid at generated position 5):
  prompt_tokens=[2, 3, 4, 5, 1],
  generated=[2, 3, 4, 5, 1, 5, 1, 2, 2, 2, 3, 3, 4, 4, 4]
traj=0 subject=279: 5/15 valid (first invalid at generated position 5):
  prompt_tokens=[9, 10, 11, 12, 13, 14, 15, 16, 17, 1],
  generated=[2, 3, 4, 5, 1, 5, 1, 2, 2, 2, 3, 3, 4, 4, 4]
traj=0 subject=287: 5/15 valid (first invalid at generated position 5):
  prompt_tokens=[6, 7, 8, 1],
  generated=[2, 3, 4, 5, 1, 5, 1, 2, 2, 2, 3, 3, 4, 4, 4]
```

Every subject and every one of the 8 trajectories emits the *same* token stream despite different prompts, goes invalid at position 5, and degenerates into runs of repeated tokens.

On the PR base the same run produces the same shape of failure (tail differs slightly: `[2,3,4,5,1,5,1,1,2,2,2,3,3,3,4]`), 128/128 invalid, first invalid at position 5.

## What it is *not*

`SGLangBackend.generate_chunk` is faithful to HF. Measured directly against a live engine on a 4-layer Llama (`head_dim=64`, vocab 3954), greedy, EOS disabled:

- **Uniform unpadded prompts, k=12 new tokens:** `SGLang(k) == HF(k)` on 3/3 rows, exactly.
- **One token at a time, appending as the rolling loop does:** `SGLang(1)×12 == HF(12)` on 3/3 rows, exactly. So the small per-chunk budgets that #168's ceiling forces are not the problem.
- **Left-padded, variable-length batch (real lengths 4/5/10/7) with an attention mask** — the shape the CLI always produces: `SGLang == HF` on 4/4 rows, exactly. So `_strip_padding_to_lists` and the mask handling are not the problem.

Nor is it fp16-vs-fp32 drift in any obvious way: those exact-match results are HF fp32 against SGLang's downcast fp16.

## Where I stopped

I did not isolate the root cause. Given per-chunk parity is exact on every shape I could construct, the divergence appears to come from the rolling loop's *interaction* with the backend rather than from a single `generate_chunk` call — the accumulation, re-priming, or per-row termination bookkeeping across chunks. That is where I would look next.

Worth noting the grammar fixture is an extreme regime (`max_seq_len=16`, ~18-token vocabulary, 400-epoch CPU pretrain), so a first step might be reproducing at a larger `max_seq_len` to see whether the effect persists or is specific to very short windows.

## Why it matters

Rolling generation is the documented path for trajectories longer than the context window (README §3.1, and the whole point of #82). Until this is understood, SGLang-backed trajectories from that path shouldn't be treated as interchangeable with HF's — the single-chunk path looks fine, but the rolling path does not.

**Comment by mmcdermott** [comment]
**Root-caused and fixed in #172.** Verified on real GPU hardware; `tests/grammar/test_cli_sglang.py` now passes with default settings and SGLang's output is byte-identical to HF's (0/128 grammar-invalid, was 128/128).

**Cause:** SGLang reads at most `context_len - 7` prompt tokens and silently drops the tail beyond that:

```
max_req_len       = min(context_len - 1, kv_pool - 1)   # tp_worker.py:288
max_req_input_len = max_req_len - 5                     # tp_worker.py:291
if len(input_ids) >= max_req_input_len: truncate        # managers/utils.py:111
```

`SGLangBackend` forces `allow_auto_truncate=True`, which turns SGLang's rejection into a truncation, and SGLang's own truncation warning is emitted through a logger its engine init has already set to ERROR — so nothing surfaces. `_rolling_generate` defaults its window to `max_seq_len - 1`, well above the ceiling, so every rolling chunk was conditioned on a truncated prefix.

Ruled out along the way, each with a measurement: greedy-vs-sampling (greedy was correct throughout), the attention backend (`triton` and `torch_native` fail identically; `head_dim` 32 vs 64 makes no difference), the radix cache (disabled makes no difference), and `generate_chunk` fidelity (exact HF parity below the ceiling for uniform, one-token-at-a-time, and left-padded variable-length batches).
